pkg/controller/buildrun/buildrun_controller.go

```diff
@@ -130,17 +131,37 @@ func (r *ReconcileBuildRun) Reconcile(request reconcile.Request) (reconcile.Resu
 		return reconcile.Result{}, nil
 	}
 
-	// Add creds to service account
-	serviceAccount, err := r.retrieveServiceAccount(buildRun)
-	if err != nil {
-		reqLogger.Error(err, "Failed to get ServiceAccount from BuildRun", "BuildRun", buildRun.Name)
-		return reconcile.Result{}, err
-	}
-	serviceAccount = applyCredentials(build, serviceAccount)
-	err = r.client.Update(context.TODO(), serviceAccount)
-	if err != nil {
-		reqLogger.Error(err, "Failed to update ServiceAccount", "ServiceAccount", serviceAccount.Name)
-		return reconcile.Result{}, err
+	// Choose a service account to use
+	serviceAccount := &corev1.ServiceAccount{}
+	if buildRun.Spec.ServiceAccount != nil && buildRun.Spec.ServiceAccount.Generate == true {
+		serviceAccount.Name = build.Name + "-sa"
+		serviceAccount.Namespace = build.Namespace
+		serviceAccount.Labels = map[string]string{buildv1alpha1.LabelBuild: build.Name,}
+		ownerReferences := metav1.NewControllerRef(build, buildv1alpha1.SchemeGroupVersion.WithKind("Build"))
+		serviceAccount.OwnerReferences = append(serviceAccount.OwnerReferences, *ownerReferences)
+
+		// Add credentials to service account
+		serviceAccount = applyCredentials(build, serviceAccount)
+		err = r.client.Create(context.TODO(), serviceAccount)
+		if err != nil {
+			reqLogger.Error(err, "Failed to create new ServiceAccount", "ServiceAccount", serviceAccount.Name)
+			return reconcile.Result{}, err
+		}
+
+	} else {
+		serviceAccount, err = r.retrieveServiceAccount(buildRun)
```

Comment:
It is possible for spec.serviceAccount.Name to be an empty string? Do we want to distinguish between empty string and skipped field by making it a 
```
// +optional
Name *string `json:"name,omitempty"`
```

Reply:
Hi @sbose78,

Thanks for the point. Yes, I think I missed one case.

Once the name of ServiceAccount is empty, we need to set it to the default serviceaccount.

I improve in the second commit.

Reply:
```
	if buildRun.Spec.ServiceAccount == nil || buildRun.Spec.ServiceAccount.Name == nil {
```

I see this check in two places, incl. `retrieveServiceAccount(..)` . Anyway we could consolidate this?

Reply:
<img width="595" alt="截屏2020-04-17下午9 37 51" src="https://user-images.githubusercontent.com/18108042/79575043-b6c51900-80f3-11ea-8791-94c5293b7132.png">

Hi @sbose78 ,

There are only two places to check the `ServiceAccount`, but they are different conditions and used in two different cases.

- The first one is used to check if the generate is `true` and auto generate the serviceaccount.

- The second one is inside of `retrieveServiceAccount` func, although it also checks the `buildRun.Spec.ServiceAccount`, but I think this func may be used in future, so it is better to keep the whole check logic inside of func, so that other logic can leverage as normal.

So if you are talking about the two checks in above screenshot, I suggest to keep them :)

What do you think?

Reply:
I was wondering if we could make `r.retrieveServiceAccount(..)` do everything? 
* see if something has been specified
* see if we've been asked to generate
* return the right one

Reply:
Hi @sbose78 ,

I got your point. Good idea. I did follow improvement and also combine the PR https://github.com/redhat-developer/build/pull/112, because there are really many overlapping between two PRs:

- Moved all service account logic into the func `retrieveServiceAccount`, I created a list to show the condtion logic covers all situations:

|ServiceAccount|Name|Generate|Result|Condition|
|---|---|---|---|---|
|Nil|Nil|Nil|pipeline sa|Condtion 2: else{ if buildRun.Spec.ServiceAccount == nil || buildRun.Spec.ServiceAccount.Name == nil|
|Set|Nil|Nil|pipeline sa|Condtion 2: else{ if buildRun.Spec.ServiceAccount == nil || buildRun.Spec.ServiceAccount.Name == nil|
|Set|Set|Nil|set sa|Condtion 3: else{ else|
|Set|Set|true|auto-generate|Condtion 1: if buildRun.Spec.ServiceAccount != nil && buildRun.Spec.ServiceAccount.Generate == true|
|Set|Set|false|set name|Condtion 3: else{ else|
|Set|Nil|true|auto-generate|Condtion 1: if buildRun.Spec.ServiceAccount != nil && buildRun.Spec.ServiceAccount.Generate == true|
|Set|Nil|false|pipeline sa|Condtion 2: else{ if buildRun.Spec.ServiceAccount == nil || buildRun.Spec.ServiceAccount.Name == nil|

- I also merged the change of https://github.com/redhat-developer/build/pull/112
To cover the bug in that PRtogether

- Yes, I also update the buildRun status if there is something wrong of sa, the result like this:
```
kubectl get br
NAME                        SUCCEEDED   REASON                                 STARTTIME   COMPLETIONTIME
buildpack-nodejs-buildrun   False       ServiceAccount "pipeline2" not found   23s
```

I think the basic failure status can follow like this style. Let us also discuss for other failure status in the issue: https://github.com/redhat-developer/build/issues/124

I also verified the e2e tests, it works:
```
✔ $ operator-sdk test local ./test/e2e --go-test-flags "-v -timeout=20m" --up-local --namespace default
INFO[0000] Testing operator locally.                    
time="2020-04-18T13:17:00+08:00" level=info msg="Started local operator"
=== RUN   TestBuild
=== RUN   TestBuild/build-group
=== RUN   TestBuild/build-group/Build_e2e_tests
=== PAUSE TestBuild/build-group/Build_e2e_tests
=== CONT  TestBuild/build-group/Build_e2e_tests
--- PASS: TestBuild (770.22s)
    --- PASS: TestBuild/build-group (0.00s)
        --- PASS: TestBuild/build-group/Build_e2e_tests (638.00s)
            main_test.go:69: Initialized cluster resources
            wait_util.go:48: Operator is running locally; skip waitForDeployment
            client.go:62: resource type ClusterBuildStrategy with namespace/name (/kaniko) created
            client.go:62: resource type Build with namespace/name (default/example-build-kaniko) created
            client.go:62: resource type BuildRun with namespace/name (default/example-build-kaniko) created
            client.go:62: resource type ClusterBuildStrategy with namespace/name (/kaniko) created
            client.go:62: resource type Build with namespace/name (default/kaniko-custom-context-dockerfile) created
            client.go:62: resource type BuildRun with namespace/name (default/kaniko-custom-context-dockerfile) created
            client.go:62: resource type ClusterBuildStrategy with namespace/name (/source-to-image) created
            client.go:62: resource type Build with namespace/name (default/example-build-s2i) created
            client.go:62: resource type BuildRun with namespace/name (default/example-build-s2i) created
            client.go:62: resource type ClusterBuildStrategy with namespace/name (/buildah) created
            client.go:62: resource type Build with namespace/name (default/example-build-buildah) created
            client.go:62: resource type BuildRun with namespace/name (default/example-build-buildah) created
            client.go:62: resource type ClusterBuildStrategy with namespace/name (/buildah) created
            client.go:62: resource type Build with namespace/name (default/buildah-custom-context-dockerfile) created
            client.go:62: resource type BuildRun with namespace/name (default/buildah-custom-context-dockerfile) created
            client.go:62: resource type ClusterBuildStrategy with namespace/name (/buildpacks-v3) created
            client.go:62: resource type Build with namespace/name (default/example-build-buildpacks-v3) created
            client.go:62: resource type BuildRun with namespace/name (default/example-build-buildpacks-v3) created
            client.go:62: resource type BuildStrategy with namespace/name (default/buildpacks-v3) created
            client.go:62: resource type Build with namespace/name (default/example-build-buildpacks-v3-namespaced) created
            client.go:62: resource type BuildRun with namespace/name (default/example-build-buildpacks-v3-namespaced) created
PASS
...
ok      github.com/redhat-developer/build/test/e2e      808.498s
INFO[0811] Local operator test successfully completed.  
```

Pls review.

Reply:
Thanks! While I review the PR, could you please update the PR title & description?

Reply:
Sure thing. Already updated the PR title and description.